dynamiqs/qarray.py

```diff
@@ -0,0 +1,558 @@
+        """
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, ...]:
+        """Returns the shape of the quantum state.
+
+        Returns:
+            The shape of the quantum state.
+        """
+
+    @property
+    def ndim(self) -> int:
+        """Returns the number of dimensions of the quantum state.
+
+        Returns:
+            The number of dimensions of the quantum state.
+        """
+        return len(self.dims)
+
+    @property
+    @abstractmethod
+    def I(self) -> QArray:  # noqa: E743
+        """Returns the identity operator compatible with the quantum state.
+
+        Returns:
+            The identity operator.
+        """
+
+    @property
+    @abstractmethod
+    def mT(self) -> QArray:
+        """Returns the transpose of the quantum state.
+
+        Returns:
+            The transpose of the quantum state.
+        """
+
+    # methods
+    @abstractmethod
+    def conj(self) -> QArray:
+        """Returns the conjugate of the quantum state.
+
+        Returns:
+            The conjugate of the quantum state.
+        """
+
+    @abstractmethod
+    def norm(self) -> Array:
+        """Returns the norm of the quantum state.
+
+        Returns:
+            The norm of the quantum state.
+        """
+
+    @abstractmethod
+    def unit(self) -> QArray:
+        """Returns the normalize the quantum state.
+
+        Returns:
+            The normalized quantum state.
+        """
+
+    @abstractmethod
+    def diag(self) -> Array:
+        """Returns the diagonal of the quantum state.
+
+        Returns:
+            The diagonal of the quantum state.
+        """
+
+    @abstractmethod
+    def trace(self) -> Array:
+        """Returns the trace of the quantum state.
+
+        Returns:
+            The trace of the quantum state.
+        """
+
+    @abstractmethod
+    def reshape(self, *new_shape: int) -> QArray:
+        """Returns the reshape the quantum state.
+
+        Args:
+            new_shape: New shape of the quantum state.
+
+        Returns:
+            The reshaped quantum state.
+        """
+
+    @abstractmethod
+    def broadcast_to(self, *new_shape: int) -> QArray:
+        """Returns the broadcast the quantum state.
+
+        Returns:
+            The broadcast quantum state.
+        """
+
+    @abstractmethod
+    def ptrace(self, keep: tuple[int, ...]) -> QArray:
+        """Returns the partial trace of the quantum state.
+
+        Args:
+            keep: Dimensions to keep.
+
+        Returns:
+            The partial trace of the quantum state.
+        """
+
+    @abstractmethod
+    def dag(self) -> QArray:
+        """Returns the dagger of the quantum state.
+
+        Returns:
+            The dagger of the quantum state.
+        """
+
+    @abstractmethod
+    def isket(self) -> bool:
+        """Returns the check if the quantum state is a ket.
+
+        Returns:
+            True if the quantum state is a ket, False otherwise.
+        """
```

Comment:
All these functions checking the shape and most conversion methods (toket, tobra, ...) can be implemented by the base class.